server/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -2507,7 +2506,27 @@ public void acquireAllPrimaryOperationsPermits(final ActionListener<Releasable>
         verifyNotClosed();
         assert shardRouting.primary() : "acquireAllPrimaryOperationsPermits should only be called on primary shard: " + shardRouting;
 
-        asyncBlockOperations(onPermitAcquired, timeout.duration(), timeout.timeUnit());
+        asyncBlockOperations(wrapPrimaryOperationPermitListener(onPermitAcquired), timeout.duration(), timeout.timeUnit());
```

Comment:
This change verifies the primary mode after acquiring permits. I think we should verify the primary mode before acquiring permits?

Reply:
We have to verify after otherwise our check could be invalidated by the time that we acquire the permit.

Reply:
ok, I missed the relocation situation. Thanks for the explanation.

Reply:
No problem. 😄 

LGTM to you then?